### DISCUSSION TITLE
Recognizing cameras connected via POE switch

### DISCUSSION THREAD

**Discussion by sunmart** [bonsai-rx/bonsai#1211] [Q&A]
Hi everyone,

I've been trying to set up behavioral data acquisition in my lab wherein we sync ultrasonic vocalization and video inputs. We're currently using a POE switch to connect 9 GeoVision cameras for simultaneous recordings. However, I'm having trouble trying to get the VideoDeviceCapture node to recognize the cameras either because they're not directly connected to the computer and/or because the GeoVision acquisition system requires us to manually convert our files to avi after acquisition. Is there a way we can have an ethernet source node that is connected to a VideoWriter node to solve this issue?

Thanks for your help!

**Comment by bruno-f-cruz** [comment]
Heya,
Never used these cameras, so no idea how they work and interface with windows.

- If the accompanying software has some way of emulating a DirectShow camera driver, it might be the easiest option indeed.

- If they establish an IP stream, you can try to access it directly in Bonsai using the `MjpegStream`.
Hope it helps,
Bruno

**Comment by sunmart** [comment]
Hi Bruno,

We couldn't get the geovision cameras to work so we got a Basler system
instead. I've been trying to figure out a way to get timestamps for video
and audio separately for analysis (ie, we want to be able to analyse the
observed mouse behavior and the corresponding audio using different
pipelines, but we want to correlate the two so we need precise timestamps).
I've attached the workflow I've been trying to do this with. Back when I
was setting it up, I used a normal webcam and I could get the frame with
the timestamp for the video. Since using PylonCapture, the timestamp node
only outputs timestamps without the frame ID. I saw from a previous post
that you said the output should be *Timestamped<IplImage>, *but that isn't
in the options for outputs I see. Do you have any suggestions?

Thanks so much for your help!
Sunaina

**Reply by bruno-f-cruz** [reply]
I don't think there's an attached workflow.
I am also not sure in what context I wrote that down. You can get different kinds of timestamps using PylonCapture. You can get the hardware timestamp using the GrabResult field inside the output of the PylonCapture, or you can timestamp frames in software and obtain a Timestamped<> object. If you are aligning to a parallel audio stream, you might want to consider what level of precision you need. If you want alignment at the hardware level, you might want to consider injecting the same signal in both streams simultaneously

**Reply by sunmart** [reply]
Oh sorry about that. I'm unable to attach the file for some reason. Below is the workflow I have at the moment. Additionally, I'm only able to save videos under 5 mins long after which the videos return an "unsupported format" error.  I saw from a pervious post on here that changing the size to 2048x2048 should solve the issue but it didn't for me. I'm not using an ffmpeg VideoWriter node though so unsure if that's the problem.

Sorry if these are really basic questions, I'm just getting started with bonsai. Thanks so much for your help!!

<img width="302" alt="bonsai workflow" src="https://user-images.githubusercontent.com/91509235/212566006-963816c7-b7d2-4d26-9a8f-0333356f898e.PNG">

**Reply by bruno-f-cruz** [reply]
You need to package your workflow inside a .zip file. Also, what version of bonsai are you using? Is it 32 or 64 bits?

**Reply by sunmart** [reply]
I'm using Bonsai 2.7.2. It is 32 bits. Here's the zip file - 
[bonsai workflow.zip](https://github.com/bonsai-rx/bonsai/files/10421234/bonsai.workflow.zip)

**Reply by bruno-f-cruz** [reply]
Can you try the 64bit version and check if it crashes?

**Reply by sunmart** [reply]
I tried the 64 bit version and it crashed.

**Reply by bruno-f-cruz** [reply]
With the same exact error?

**Reply by sunmart** [reply]
No error, the program just stopped and exited.

**Reply by bruno-f-cruz** [reply]
> I'm using Bonsai 2.7.2. It is 32 bits. Here's the zip file - [bonsai workflow.zip](https://github.com/bonsai-rx/bonsai/files/10421234/bonsai.workflow.zip)

Can't open this zip file. Can you try again? [Maybe this would help?](https://support.microsoft.com/en-us/windows/zip-and-unzip-files-8d28fa72-f2f9-712f-67df-f80cf89fd4e5)

**Comment by sunmart** [comment]
[bonsai workflow.zip](https://github.com/bonsai-rx/bonsai/files/10421620/bonsai.workflow.zip)

**Reply by bruno-f-cruz** [reply]
Looking at the workflow I am not sure why it is crashing when trying to save. That said, you should use the 64bit version, especially when saving larger videos.

Regarding synchronization, I am not sure what you are trying to achieve.  If you want a `Timestamped<IplImage>` you can simply add the timestamp after the `Image`. My question would be if you are ok with synchronizing everything in software or whether you need higher precision in your experiment.

**Comment by sunmart** [comment]
Hi Bruno,

I uninstalled Bonsai and downloaded it back again and I was able to solve the issue with it crashing when trying to save videos longer than 6 mins using the 64 bit version. That being said, 10 min video recordings are yielding 3 min videos. Since I was able to get 6 min recordings before, it might be that the frames are getting overwritten at some point.  I'm using an 8-core processor with 64GB RAM. I reduced the size of the frames to 1280x720 and cropped the field of view and that solved the issue of dropped frames on Pylon viewer but not in Bonsai. 

Since we're trying to correlate behavior epochs with USVs which occur over seconds, I think synchronizing in software should be okay provided there is little information loss. But given that I'm having issues with dropped frames, it might be good to know how to do it with higher precision. 

Thanks again for all your help!

**Reply by bruno-f-cruz** [reply]
How have you diagnosed the dropped frames? Did you count them? Are you triggering each frame?

**Reply by sunmart** [reply]
Pylon viewer has a counter for dropped frames so after I saw that the videos were short, I looked at the frame count during acquisition on Pylon viewer. I am not triggering each frame.

**Comment by bruno-f-cruz** [comment]
Re: dropping frames I find it odd you are dropping so many. I would
probably try to see what is going on. Probably start counting them in the
output video. Making sure the usb connection is stable. Lowering the data
volume. Just to list a few.

Regarding synchrony, if you are ok with some jitter then simply
timestamping the buffers in bonsai should be enough. If you want precise
alignment, it kinda depends on your setup. What kind of setup are you using
to record audio?

**Reply by sunmart** [reply]
Hi Bruno,

Apologies for the delayed response, I was trying to figure out how to fix the problem of the dropped frames. I'm now acquiring video directly from the camera software and using an arduino serial monitor to input frame numbers into bonsai which I'm timestamping. 

I would prefer precise alignment just because we intend to do photometry eventually. We have two pettersson ultrasound microphones that are connected directly to a USB bus that I'm feeding into Bonsai for timestamping.  I've attached the workflow here. Maybe this will be better. 
[Audio_arduino_sync.zip](https://github.com/bonsai-rx/bonsai/files/10494955/Audio_arduino_sync.zip)

**Reply by bruno-f-cruz** [reply]
I have never used these mics so take the following with a (large) grain of salt...

Since you are already triggering the camera on each frame, if you want to align frames to audio (and the two audio streams too for that matter) you could just duplicate the TTL that drives the camera to the audio card, and use one of the input channels to record it. Not sure what the input voltage tolerances of the ADC of the audio card are, but assuming that you can get it to work, you would essentially have as good alignment as you possibly could. 

Otherwise, I would probably try to benchmark your timestamps. From past experience, you might end up with something with a jitter of +-10ms on each buffer you receive from the audio card with fairly low latency (not sure with USB audio cards). Similar jitter values for the Arduino, albeit with a larger delay (~5-10ms if I recall correctly).
